include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+{
+	//vec2 thickness_eta = unpackHalf2x16(data.data[3].x);
+	//vec3 sigmaA = params_getSigmaA(params);
+	//float a = params_getAlpha(params);
+	return bxdf_eval_t(1.0,0.0,0.0);
+}
+
+void instr_execute_BUMPMAP(in instr_t instr, in mat2x4 srcs)
+{
+	vec3 N = srcs[0].xyz;
+	setCurrInteraction(N);
+}
+
+void instr_execute_SET_GEOM_NORMAL()
+{
+	setCurrInteraction(normalize(Normal));
+}
+
+bxdf_eval_t instr_execute_cos_eval_BLEND(in instr_t instr, in mat2x4 srcs, in params_t params, in bsdf_data_t data)
+{
+	float w = params_getBlendWeight(params);
+	bxdf_eval_t bxdf1 = srcs[0].xyz;
+	bxdf_eval_t bxdf2 = srcs[1].xyz;
+
+	bxdf_eval_t blend = mix(bxdf1, bxdf2, w);
+	return blend;
+}
+eval_and_pdf_t instr_execute_cos_eval_pdf_BLEND(in instr_t instr, in mat2x4 srcs, in params_t params, in bsdf_data_t data)
+{
+	float w = params_getBlendWeight(params);
+	eval_and_pdf_t bxdf1 = srcs[0];
+	eval_and_pdf_t bxdf2 = srcs[1];
+
+	float wa = w;
+	float wb = 1.0-wa;
+	bxdf_eval_t a = bxdf1.rgb;
+	float pdfa = bxdf1.a;
+	bxdf_eval_t b = bxdf2.rgb;
+	float pdfb = bxdf2.a;
+	float pdf = mix(pdfa,pdfb,wa);
+
+	bxdf_eval_t blend;
+	//instead of doing this here, remainder and pdf returned from generator stream is weighted
+	//so it correctly adds up at the end
+	/*
+	//generator is denoted as the one with negative probability
+	if (pdfa<0.0) {//`a` is remainder
+		pdfa = abs(pdfa);
+		vec3 rem = (a*wa + b/pdfa*wb) / (wa + pdfb/pdfa*wb);
+		blend = rem_and_pdf_t(rem, pdf);
+	}
+	else
+	if (pdfb<0.0) {//`b` is remainder
+		pdfb = abs(pdfb);
+		vec3 rem = (b*wb + a/pdfb*wa) / (wb + pdfa/pdfb*wa);
+		blend = rem_and_pdf_t(rem, pdf);
+	}
+	else {
+		vec3 rem = mix(a,b,wa);
+		blend = rem_and_pdf_t( rem,pdf );
+	}
+	*/
+	blend = mix(a,b,wa);
+	
+	return eval_and_pdf_t(blend, pdf);
+}
+
+vec3 fetchTex(in uvec3 texid, in vec2 uv, in mat2 dUV)
+{
+	float scale = uintBitsToFloat(texid.z);
+
+	return irr_glsl_vTextureGrad(texid.xy, uv, dUV).rgb*scale;
+}
+
+uint instr_getTexFetchFlags(in instr_t i)
+{
+	uint flags = bitfieldExtract(i.x,INSTR_FETCH_FLAG_TEX_0_SHIFT,1);
+	flags |= bitfieldExtract(i.x,INSTR_FETCH_FLAG_TEX_1_SHIFT,1)<<1;
+	flags |= bitfieldExtract(i.x,INSTR_FETCH_FLAG_TEX_2_SHIFT,1)<<2;
+	flags |= bitfieldExtract(i.x,INSTR_FETCH_FLAG_TEX_3_SHIFT,1)<<3;
+
+	return flags;
+}
+
+uvec4 instr_decodePrefetchRegs(in instr_t i)
+{
+	return 
+	(i.yyyy >> (uvec4(INSTR_PREFETCH_REG_0_SHIFT,INSTR_PREFETCH_REG_1_SHIFT,INSTR_PREFETCH_REG_2_SHIFT,INSTR_PREFETCH_REG_3_SHIFT)-32u)) & uvec4(INSTR_PREFETCH_REG_MASK);
+}
+
+void runTexPrefetchStream(in instr_stream_t stream, in mat2 dUV)
+{
+	for (uint i = 0u; i < stream.count; ++i)
+	{
+		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
+
+		uvec4 regcnt = 
+		( instr.yyyy >> (uvec4(INSTR_FETCH_TEX_0_REG_CNT_SHIFT,INSTR_FETCH_TEX_1_REG_CNT_SHIFT,INSTR_FETCH_TEX_2_REG_CNT_SHIFT,INSTR_FETCH_TEX_3_REG_CNT_SHIFT)-uvec4(32)) ) & uvec4(INSTR_FETCH_TEX_REG_CNT_MASK);
+		bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
```

Comment:
this is messed up... it increases divergence in a clandestine way

There's a loop of a variable number of prefetch instructions, each of which can have a variable number of texture fetches (0-4).

Why can't we preprocess this off-line and unroll the current prefetch instrction + bsdf_data texture handles into a simple texture prefetch stream where each instruction will prefetch only one texture/constant and write to only one set of registers?

Also doing this offline on CPU (preparing the list), would allow us to thin out the BSDF data (remove all textures and constants that would be prefetched)

Reply:
a prefetch instruction would need to be a `uvec4` instead of a `uvec2`